spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/r2dbc/ConnectionFactoryBuilder.java

```diff
@@ -58,8 +62,26 @@ private ConnectionFactoryBuilder(ConnectionFactoryOptions.Builder optionsBuilder
 	 */
 	public static ConnectionFactoryBuilder of(R2dbcProperties properties,
 			Supplier<EmbeddedDatabaseConnection> embeddedDatabaseConnection) {
-		return new ConnectionFactoryBuilder(
-				new ConnectionFactoryOptionsInitializer().initializeOptions(properties, embeddedDatabaseConnection));
+		return of(properties, embeddedDatabaseConnection, ProxyListenersProvider.EMPTY);
+	}
+
+	/**
+	 * Initialize a new {@link ConnectionFactoryBuilder} based on the specified
+	 * {@link R2dbcProperties}. If no url is specified, the
+	 * {@link EmbeddedDatabaseConnection} supplier is invoked to determine if an embedded
+	 * database can be configured instead.
+	 * @param properties the properties to use to initialize the builder
+	 * @param embeddedDatabaseConnection a supplier for an
+	 * {@link EmbeddedDatabaseConnection}
+	 * @param proxyListenersProvider a supplier for list of {@link ProxyExecutionListener}
+	 * @return a new builder initialized with the settings defined in
+	 * {@link R2dbcProperties}
+	 */
+	public static ConnectionFactoryBuilder of(R2dbcProperties properties,
+			Supplier<EmbeddedDatabaseConnection> embeddedDatabaseConnection,
+			ProxyListenersProvider proxyListenersProvider) {
+		return new ConnectionFactoryBuilder(new ConnectionFactoryOptionsInitializer().initializeOptions(properties,
+				embeddedDatabaseConnection, proxyListenersProvider));
```

Comment:
I wouldn't surface this so high. If post-processing the `ConnectionFactory` and returning another instance is something that's required by `r2dbc-proxy`, perhaps the builder should expose such hook-point and offer a method on the builder to register such implementation of that hook-point.